### PR TITLE
refactor: re-implement desktop-capturer in TypeScript

### PR DESCRIPTION
**Pull request by miniak**
#### Description of Change
Rewrite in TypeScript and simplify the implementation by using async / await.

#### Checklist
- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)

#### Release Notes
Notes: no-notes

### CODE REVIEW

File: lib/browser/desktop-capturer.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+      resolve(sources.map(source => ({
+        id: source.id,
+        name: source.name,
+        thumbnail: source.thumbnail.toDataURL(),
+        display_id: source.display_id,
+        appIcon: (fetchWindowIcons && source.appIcon) ? source.appIcon.toDataURL() : null
+      })))
+    })
+
+    let capturer: NodeJS.DesktopCapturer | null = createDesktopCapturer()
```
[the remark this answers is not in the document]
Comment (miniak) [reply lib/browser/desktop-capturer.ts]: this is the native binding, those are in the NodeJS namespace. it's different than the public API

[the remark this answers is not in the document]
Comment (miniak) [reply lib/browser/desktop-capturer.ts]: @MarshallOfSound should I move it to `ElectronInternal` instead maybe?

[the remark this answers is not in the document]
Comment (codebytere) [reply lib/browser/desktop-capturer.ts]: @miniak right, sorry, that's what I meant: that this should be moved _out_ of the NodeJS namespace.

File: typings/internal-electron.d.ts [hunk 2]
```diff
@@ -47,11 +47,6 @@ declare namespace Electron {
     allFrames: boolean
   }
 
-  interface RendererProcessPreference {
```
Comment (miniak) [inline typings/internal-electron.d.ts]: leftover from #18532

File: lib/browser/desktop-capturer.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (miniak) [reply lib/browser/desktop-capturer.ts]: done

**Pull request merged**